data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/ScanObjectWorker.java

```diff
@@ -198,8 +231,141 @@ private Optional<DeleteObjectRequest> processS3Object(final S3ObjectReference s3
         return Optional.empty();
     }
 
+    private void processFolderPartition(final SourcePartition<S3SourceProgressState> folderPartition) {
+        final String bucket = folderPartition.getPartitionKey().split("\\|")[0];
+        final String s3Prefix = folderPartition.getPartitionKey().split("\\|")[1];
+
+        final List<S3ObjectReference> objectsToProcess = getObjectsForPrefix(bucket, s3Prefix);
+
+        if (objectsToProcess.isEmpty()) {
+            partitionKeys.remove(folderPartition.getPartitionKey());
+            if (shouldDeleteFolderPartition(folderPartition)) {
+                LOG.info("Deleting folder partition {} as no objects have been found from this folder for {} minutes", folderPartition.getPartitionKey(), NO_OBJECTS_FOUND_BEFORE_PARTITION_DELETION_DURATION.toMinutes());
+                sourceCoordinator.deletePartition(folderPartition.getPartitionKey());
+                return;
+            }
+
+            if (folderPartition.getPartitionState().isEmpty()) {
+                final S3SourceProgressState s3SourceProgressState = new S3SourceProgressState(Instant.now().toEpochMilli());
+                sourceCoordinator.saveProgressStateForPartition(folderPartition.getPartitionKey(), s3SourceProgressState);
+            }
+
+            sourceCoordinator.giveUpPartition(folderPartition.getPartitionKey(), Instant.now());
+            return;
+        }
+
+        Optional<S3SourceProgressState> folderPartitionProgressState = folderPartition.getPartitionState();
+        if (folderPartitionProgressState.isEmpty()) {
+            folderPartitionProgressState = Optional.of(new S3SourceProgressState(Instant.now().toEpochMilli()));
+        } else {
+            folderPartitionProgressState.get().setLastTimeObjectsFound(Instant.now().toEpochMilli());
+        }
+
+        // Update the last time objects were found to support deletion of the partition after no objects found for some time
+        sourceCoordinator.saveProgressStateForPartition(folderPartition.getPartitionKey(), folderPartitionProgressState.get());
+
+
+        int objectsProcessed = 0;
+        int objectIndex = 0;
+
+        AcknowledgementSet acknowledgementSet = null;
+        String activeAcknowledgmentSetId = null;
+        while (objectIndex < objectsToProcess.size() && objectsProcessed < folderPartitioningOptions.getMaxObjectsPerOwnership()) {
```

Comment:
This method can be split into smaller methods for readability. We can move this while loop to different method.